README.md
<h1 align="center">Reading Time Plugin for <a href="https://flextype.org/">Flextype</a></h1>

<p align="center">
<a href="https://github.com/flextype-plugins/reading-time/releases"><img alt="Version" src="https://img.shields.io/github/release/flextype-plugins/reading-time.svg?label=version&color=black"></a> <a href="https://github.com/flextype-plugins/reading-time"><img src="https://img.shields.io/badge/license-MIT-blue.svg?color=black" alt="License"></a> <a href="https://github.com/flextype-plugins/reading-time"><img src="https://img.shields.io/github/downloads/flextype-plugins/reading-time/total.svg?color=black" alt="Total downloads"></a> <a href="https://github.com/flextype/flextype"><img src="https://img.shields.io/badge/Flextype-0.9.16-green.svg?color=black" alt="Flextype"></a> <a href=""><img src="https://img.shields.io/discord/423097982498635778.svg?logo=discord&color=black&label=Discord%20Chat" alt="Discord"></a>
</p>

Reading Time plugin will help you to know how many readers of every page in your website.

## Dependencies

The following dependencies need to be installed for Reading Time Plugin.

| Item | Version | Download |
|---|---|---|
| [flextype](https://github.com/flextype/flextype) | 0.9.16 | [download](https://github.com/flextype/flextype/releases) |
| [twig](https://github.com/flextype-plugins/twig) | >=2.0.0 | [download](https://github.com/flextype-plugins/twig/releases) |

## Installation

1. Download & Install all required dependencies.
2. Create new folder `/project/plugins/reading-time`
3. Download Reading Time Plugin and unzip plugin content to the folder `/project/plugins/reading-time`

## Settings

| Key | Value | Description |
|---|---|---|
| enabled | true | true or false to disable the plugin |
| priority | 100 | Reading Time plugin priority |

## Documentation

Function `readingTime()` in namespace `Flextype\Plugin\ReadingTime`

```php
/**
 * Get reading time
 *
 * @return string
 */
 function readingTime(string $content, array $options = []): string
```

### Usage

Display reading time with custom translates in the twig templates

```
{{ readingTime(entry.content, {
                'minute': 'Minute',
                'minutes': 'Minutes',
                'second':  'Second',
                'seconds': 'Seconds'
                }) }}
```

Display reading time with custom translates and format in the twig templates

```
{{ readingTime(entry.content, {
                'minute': 'Minute',
                'minutes': 'Minutes',
                'second':  'Second',
                'seconds': 'Seconds',
                'format': '{minutes_count} {minutes_label} – {seconds_count} {seconds_label}'
                }) }}
```

Display reading time with custom translates and format that alternative format that hides the minute label in the twig templates

```
{{ readingTime(entry.content, {
                'minute': 'Minute',
                'minutes': 'Minutes',
                'second':  'Second',
                'seconds': 'Seconds',
                'format': '{minutes_count} {minutes_label} – {seconds_count} {seconds_label}'
                }) }}
```

Display page view counter in the php

```php
echo Flextype\Plugin\ReadingTime\readingTime($content, $options);
```

## LICENSE
[The MIT License (MIT)](https://github.com/flextype-plugins/reading-time/blob/master/LICENSE.txt)
Copyright (c) 2021 [Sergey Romanenko](https://github.com/Awilum)


CHANGELOG.md
# v2.0.0, 2021-01-16
* Total plugin rewrite

# v1.3.0, 2019-12-30
* Updates for Flextype 0.9.6

# v1.2.0, 2019-10-02
* Updates for Flextype 0.9.5

# v1.1.2, 2019-01-17
* Updates for Flextype 0.8.3

# v1.1.1, 2018-11-20
* Updates for Flextype 0.7.0

# v1.1.0, 2018-05-07
* Updates for Flextype 0.3.0

# v1.0.1, 2018-03-21
* Rebranding

# v1.0.0, 2018-03-16
* Initial Release


reading-time.php
<?php namespace Rawilum;

/**
 *
 * Reading Time Plugin for Rawilum
 *
 * @author Romanenko Sergey / Awilum <awilum@yandex.ru>
 * @link http://rawilum.org
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

 function readingTime(string $content, array $params = [])
 {
     $defaults = [
       'minute'              => 'minute',
       'minutes'             => 'minutes',
       'second'              => 'second',
       'seconds'             => 'seconds',
       'format'              => '{minutes_count} {minutes_label}, {seconds_count} {seconds_label}',
       'format.alt'          => '{seconds_count} {seconds_label}',
       'format.alt.enable'   => false
     ];

     $options      = array_merge($defaults, $params);
     $words        = str_word_count(strip_tags($content));
     $minutesCount = floor($words / 200);
     $secondsCount = floor($words % 200 / (200 / 60));
     $minutesLabel = ($minutesCount <= 1) ? $options['minute'] : $options['minutes'];
     $secondsLabel = ($secondsCount <= 1) ? $options['second'] : $options['seconds'];

     $replace      = [
       'minutes_count' => $minutesCount,
       'minutes_label' => $minutesLabel,
       'seconds_count' => $secondsCount,
       'seconds_label' => $secondsLabel,
     ];

     if ($minutesCount < 1 and $options['format.alt.enable'] === true) {
         $result = $options['format.alt'];
     } else {
         $result = $options['format'];
     }

     foreach ($replace as $key => $value) {
         $result = str_replace('{' . $key . '}', $value, $result);
     }

     return $result;
 }


twig/ReadingTimeTwigExtension.php
<?php

/**
 * @package Flextype
 *
 * @author Sergey Romanenko <hello@romanenko.digital>
 * @link http://romanenko.digital
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Flextype;

class ReadingTimeTwigExtension extends \Twig_Extension
{
    /**
     * Flextype Dependency Container
     */
    private $flextype;

    /**
     * Constructor
     */
    public function __construct($flextype)
    {
        $this->flextype = $flextype;
    }

    /**
     * Callback for twig.
     *
     * @return array
     */
    public function getFunctions()
    {
        return [
            new \Twig_SimpleFunction('readingTime', [$this, 'readingTime'])
        ];
    }

    public function readingTime(string $content, array $params = [])
    {
        return $this->flextype->ReadingTimeController->readingTime($content, $params);
    }
}


app/Controllers/ReadingTimeController.php
<?php

namespace Flextype;

use Flextype\Component\Filesystem\Filesystem;

class ReadingTimeController extends Controller {

    public function readingTime(string $content, array $params = [])
    {
        $defaults = [
          'minute'              => 'minute',
          'minutes'             => 'minutes',
          'second'              => 'second',
          'seconds'             => 'seconds',
          'format'              => '{minutes_count} {minutes_label}, {seconds_count} {seconds_label}',
          'format.alt'          => '{seconds_count} {seconds_label}',
          'format.alt.enable'   => false
        ];

        $options      = array_merge($defaults, $params);
        $words        = str_word_count(strip_tags($content));
        $minutesCount = floor($words / 200);
        $secondsCount = floor($words % 200 / (200 / 60));
        $minutesLabel = ($minutesCount <= 1) ? $options['minute'] : $options['minutes'];
        $secondsLabel = ($secondsCount <= 1) ? $options['second'] : $options['seconds'];

        $replace      = [
          'minutes_count' => $minutesCount,
          'minutes_label' => $minutesLabel,
          'seconds_count' => $secondsCount,
          'seconds_label' => $secondsLabel,
        ];

        if ($minutesCount < 1 and $options['format.alt.enable'] === true) {
            $result = $options['format.alt'];
        } else {
            $result = $options['format'];
        }

        foreach ($replace as $key => $value) {
            $result = str_replace('{' . $key . '}', $value, $result);
        }

        return $result;
    }
}


plugin.php
<?php

declare(strict_types=1);

namespace Flextype\Plugin\ReadingTime {

    use Twig\TwigFunction;

    /**
     * Get reading time
     *
     * @return string
     */
     function readingTime(string $content, array $options = []): string
     {
         $defaults = [
           'minute'              => 'minute',
           'minutes'             => 'minutes',
           'second'              => 'second',
           'seconds'             => 'seconds',
           'format'              => '{minutes_count} {minutes_label}, {seconds_count} {seconds_label}',
           'format.alt'          => '{seconds_count} {seconds_label}',
           'format.alt.enable'   => false
         ];

         $options      = array_merge($defaults, $options);
         $words        = str_word_count(strip_tags($content));
         $minutesCount = floor($words / 200);
         $secondsCount = floor($words % 200 / (200 / 60));
         $minutesLabel = ($minutesCount <= 1) ? $options['minute'] : $options['minutes'];
         $secondsLabel = ($secondsCount <= 1) ? $options['second'] : $options['seconds'];

         $replace      = [
           'minutes_count' => $minutesCount,
           'minutes_label' => $minutesLabel,
           'seconds_count' => $secondsCount,
           'seconds_label' => $secondsLabel,
         ];

         if ($minutesCount < 1 and $options['format.alt.enable'] === true) {
             $result = $options['format.alt'];
         } else {
             $result = $options['format'];
         }

         foreach ($replace as $key => $value) {
             $result = str_replace('{' . $key . '}', $value, $result);
         }

         return $result;
     }

    // Twig: {{ readingTime() }}
    flextype('twig')->addFunction(new TwigFunction('readingTime', fn (string $content, array $options = []) => readingTime($content, $options)));
}
